### ISSUE TITLE
Expose model parameters as model attributes

### ISSUE THREAD

**Issue by cdeil** [gammapy/gammapy#1665]
In both Sherpa and Astropy, model parameters are exposed as model attributes. Getting returns a `Parameter` object, setting via float works.

Here's an example with Sherpa, but Astropy behaves the same way:
```
>>> from sherpa.models import Gauss1D
>>> g = Gauss1D()
>>> g.pos
<Parameter 'pos' of model 'gauss1d'>
>>> g.pos.val
0.0
>>> g.pos = 99
>>> g.pos
<Parameter 'pos' of model 'gauss1d'>
>>> g.pos.val
99.0
>>> g.pos.val = 100
>>> g.pos
<Parameter 'pos' of model 'gauss1d'>
>>> g.pos.val
100.0
```
Not all models do this, e.g. composite models like `m = Gauss1D() + Gauss1D()` don't have the parameters exposed that way.

Now the mechanism how this is implemented in Astropy and Sherpa is very different. 

In Sherpa, parameters are set on self in init, see [here](https://github.com/sherpa/sherpa/blob/2eb91d35a015bf60cf7f1ec0c21ce92648f36d59/sherpa/models/basic.py#L490-L494).

The sane behaviour of interacting with models and parameters is achieved via `__getattr__` and `__setattr__` in the Model base class:
https://github.com/sherpa/sherpa/blob/84fb8a728f8dde70e1aea94adbedd540ca98a4db/sherpa/models/model.py#L191
Parameter and Model objects are locked down, trying to set anything non-valid on them like using a wrong parameter name results in an AttributeError:
```
>>> g.position = 99
AttributeError: 'Gauss1D' object has no attribute 'position'
```

Note how easy it is to make silent mistakes when setting parameter values. E.g. here's a bug where someone wrote `source_model.parameters.index = 1.12` which adds an attribute `index` to the `source_model.parameters` object, but has no effect, the intention was to modify the parameter value via `source_model.parameters['index'] = 1.12`.

https://github.com/gammapy/gammapy/blob/6879e06392391dc523e76398eabd489b7e4ccf67/gammapy/spectrum/tests/test_fit.py#L85

In Astropy, `Parameter` is a descriptor and model parameters are declared at the class level (see [here](http://docs.astropy.org/en/stable/modeling/new.html#a-step-by-step-definition-of-a-1-d-gaussian-model). I did not look at that in detail yet.

I think we should try to make this addition in Gammapy today, and we have to decide which solution to pursue. I realise what we achieve might not be the good and final solution, but I do think with a day or two of effort this week we can make something much better than what we have now.

I'm inclined to do a simple implementation matching what Sherpa does: add a Model base class that is locked down and implements getatttr / setattr to lock down model objects, then change all model initialisers to do what Sherpa does; in a separate PR also lock down `ParameterList` and make it's `parameters` attribute private, providing only a small controlled API to interact with it.

The alternative I thing would be to study the design in Astropy, and to either adopt it directly or via subclassing, or to implement a simple version of it ourselves. There I don't think I could do this in a day, it would take more time.

@joleroi @adonath @registerrier - Thoughts?

**Comment by cdeil** [comment]
I'm postponing this to v0.9.

**Comment by adonath** [comment]
I agree this is useful addition and I think we should just adapt the solution the `sherpa` uses. It is conceptionally easier and also much easier to implement.

**Comment by adonath** [comment]
This is part of #1971, I'll move it it to v0.11 milestone, because it is not a priority right now.

**Comment by adonath** [comment]
@luca-giunti I know that you started to work on this during the coding sprint. Any plans to get this in for Gammapy v0.11? Shall we do pair coding on this this week?

**Comment by luca-giunti** [comment]
@adonath yes sure! I am in the process of implementing it, based on the sherpa example. I could do pair coding tomorrow (all day) or thursday morning, as you want

**Comment by adonath** [comment]
Implemented in #2072.
